include/realtime_tools/lock_free_queue.hpp

```diff
@@ -0,0 +1,358 @@
+   */
+
+  [[nodiscard]] bool push(const T & data) { return data_queue_.push(data); }
+
+  /**
+   * @brief Push the data into the queue
+   * @param data Data to be pushed
+   * @return true If the data is pushed successfully
+   * @return false If the queue is full or the data could not be pushed
+   * @note This function is enabled only if the data type is convertible to the template type of the queue
+   */
+  template <typename U>
+  [[nodiscard]] std::enable_if_t<std::is_convertible_v<T, U>, bool> push(const U & data)
+  {
+    return data_queue_.push(data);
+  }
+
+  /**
+   * @brief The bounded_push function pushes the data into the queue and pops the oldest data if the queue is full
+   * @param data Data to be pushed
+   * @return true If the data is pushed successfully
+   * @return false If the data could not be pushed
+   * @note This function is enabled only if the queue is a spsc_queue and only if the data type
+   * is convertible to the template type of the queue
+   * @note To be used in a single threaded applications
+   * @warning This method might not work as expected if it is used with 2 different threads one
+   * doing bounded_push and the other doing pop. In this case, the queue is no more a single producer
+   * single consumer queue. So, the behaviour might not be as expected.
+   */
+  template <
+    typename U, bool IsSPSCQueue = is_spsc_queue<LockFreeContainer>::value,
+    typename std::enable_if_t<IsSPSCQueue, int> = 0>
+  [[nodiscard]] std::enable_if_t<std::is_convertible_v<T, U>, bool> bounded_push(const U & data)
+  {
+    if (!data_queue_.push(data)) {
+      T dummy;
+      data_queue_.pop(dummy);
+      return data_queue_.push(data);
+    }
+    return true;
+  }
+
+  /**
+   * @brief Get the latest data from the queue
+   * @param data Data to be filled with the latest data
+   * @return true If the data is filled with the latest data, false otherwise
+   * @return false If the queue is empty
+   * @note This function consumes all the data in the queue until the last data and returns the last element of the queue
+   */
+  [[nodiscard]] bool get_latest(T & data)
+  {
+    return data_queue_.consume_all([&data](const T & d) { data = d; }) > 0;
+  }
+
+  /**
+   * @brief The bounded_push function pushes the data into the queue and pops the oldest data if the queue is full
+   * @param data Data to be pushed
+   * @return true If the data is pushed successfully
+   * @return false If the data could not be pushed
+   * @note This function is enabled only if the queue is of multiple producer and multiple consumer
+   * type and only if the data type is convertible to the template type of the queue
+   * @note Can be used in a multi threaded applications
+   */
+  template <
+    typename U, bool IsSPSCQueue = is_spsc_queue<LockFreeContainer>::value,
+    typename std::enable_if_t<!IsSPSCQueue, int> = 1>
+  [[nodiscard]] std::enable_if_t<std::is_convertible_v<T, U>, bool> bounded_push(const U & data)
+  {
```

Comment:
I might be missing something, but I see this as kinda the same function in [line 203](https://github.com/ros-controls/realtime_tools/pull/253/files#diff-ea7cd06f818c3d7ad61d89fe4cee66ec3c3734c83050862a7754d01768e4e6a4R203), is there any reason to do this separately?

maybe it would be better if we handle the retries with a while loop and a predefined (configurable) limit

Reply:
Done

Reply:
Thanks for the suggestions. I applied to other functions as well